src/msw/toolbar.cpp

```diff
@@ -1895,6 +1897,52 @@ void wxToolBar::OnEraseBackground(wxEraseEvent& event)
 #endif // wxHAS_MSW_BACKGROUND_ERASE_HOOK
 }
 
+void wxToolBar::RealizeHelper()
+{
+    Realize();
+}
+
+void wxToolBar::OnDPIChanged(wxDPIChangedEvent& event)
+{
+    // Manually scale the size of the controls. Even though the font has been
+    // updated, the internal size of the controls does not.
+    const float scaleFactor = (float)event.GetNewDPI().y / event.GetOldDPI().y;
+
+    wxToolBarToolsList::compatibility_iterator node;
+    for ( node = m_tools.GetFirst(); node; node = node->GetNext() )
+    {
+        wxToolBarTool* const tool = static_cast<wxToolBarTool*>(node->GetData());
+        if ( !tool->IsControl() )
+            continue;
+
+        if ( wxControl* const control = tool->GetControl() )
+        {
+            const wxSize oldSize = control->GetSize();
+            wxSize newSize = oldSize * scaleFactor;
+
+            // choice based controls seem to automatically adjust their height
+            // when the font size increases, keep this size
+            if ( dynamic_cast<wxChoiceBase*>(control) && scaleFactor > 1 )
```

Comment:
We should use `wxDynamicCast()` here, some people disable RTTI when building wxMSW.

But beyond this, it would be nice to avoid the cast by adding some virtual `MSWAdjustSizeOnDPIChange()` that would just call `SetSize()` by default but could be overridden in `wxChoice` to only adjust the width and not the height.